src/engine/TPA.cc

```diff
@@ -1588,10 +1591,7 @@ VerificationResult TransitionSystemNetworkManager::solve() && {
     activePath.clear();
     while (true) {
         if (getNode(current).blocked_children == getNode(current).children.size()) {
-            if (current == graph.getEntry()) {
-                // TODO: Compute witness
-                return {VerificationAnswer::SAFE, NoWitness{"Unable to compute witness from TPA for DAG of TSs yet"}};
-            }
+            if (current == graph.getEntry()) { return {VerificationAnswer::SAFE, computeValidityWitness()}; }
```

Comment:
We actually need to check if we need to compute witnesses or not.
If not, there is not point in trying to compute it, it would actually be harmful.
That concerns both valid and invalid case.

Reply:
Oh, you mean specifically if there is a demand to compute witness or not...

Reply:
I don't see much harm, but slowdown is indeed guaranteed(basically if no validate option, we can return noWitness everywhere)

I would move it to the different PR though, as at is a different issue, connected not only with our algorithm/TPA, but I suppose also with other engines

Reply:
By harmful I meant that some instances might not be solved anymore because of the witness computation, even though the witness was not required.

It should be fixed here for exactly that reason. The change is simple, you just need to check if witness computation was requested and only call `computeValidityWitness` in such case.
TPA for a transition system already behaves that way.

Reply:
The thing is, this issue is uniform, it also exists for Spacer, Lawi, etc, because we always produce witnesses for SAFE/UNSAFE. Therefore imho it should be a uniform change all over the project, that is somewhat independent from this specific DAG approach. I've added a check, but I think it should be general tbh...